File: sdk/typescript/src/harness/providers/factory.ts
import type { HarnessProvider } from './base.js';
import type { HarnessConfig } from '../types.js';
import { ClaudeCodeProvider } from './claude.js';
import { CodexProvider } from './codex.js';

export const SUPPORTED_PROVIDERS = new Set(['claude-code', 'codex', 'gemini', 'opencode']);

export function buildProvider(config: HarnessConfig): HarnessProvider {
  if (!SUPPORTED_PROVIDERS.has(config.provider)) {
    throw new Error(
      `Unknown harness provider: "${config.provider}". Supported: ${[...SUPPORTED_PROVIDERS].sort().join(', ')}`
    );
  }
  if (config.provider === 'claude-code') {
    return new ClaudeCodeProvider();
  }
  if (config.provider === 'codex') {
    return new CodexProvider(config.codexBin ?? 'codex');
  }
  throw new Error(`Provider "${config.provider}" is not yet implemented.`);
}

File: sdk/python/agentfield/harness/providers/_factory.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from agentfield.harness.providers._base import HarnessProvider
    from agentfield.types import HarnessConfig

SUPPORTED_PROVIDERS = {"claude-code", "codex", "gemini", "opencode"}


def build_provider(config: "HarnessConfig") -> "HarnessProvider":
    provider_name = config.provider
    if provider_name not in SUPPORTED_PROVIDERS:
        raise ValueError(
            f"Unknown harness provider: {provider_name!r}. Supported providers: "
            f"{', '.join(sorted(SUPPORTED_PROVIDERS))}"
        )
    if provider_name == "claude-code":
        from agentfield.harness.providers.claude import ClaudeCodeProvider

        return ClaudeCodeProvider()
    if provider_name == "codex":
        from agentfield.harness.providers.codex import CodexProvider

        return CodexProvider(bin_path=getattr(config, "codex_bin", "codex"))
    raise NotImplementedError(
        f"Provider {provider_name!r} is not yet implemented. Install the corresponding provider package."
    )
